### ISSUE TITLE
[Bug] `SingleTaskGP.posterior` with multioutput evaluates the covar matrix

### ISSUE THREAD

**Issue by valtron** [pytorch/botorch#971]
# 🐛 Bug

`SingleTaskGP.posterior` with multioutput evaluates the covar matrix, making that operation O(n^2) instead of O(n) in time/space. This is not entirely a BoTorch issue: part of the problem comes from GPyTorch. More details in the "Additional context" section.

## To reproduce

```python
import torch
from botorch.models.gp_regression import SingleTaskGP

X = torch.zeros((5, 4))
Y = torch.zeros((5, 3))
gp = SingleTaskGP(X, Y)

n = 1_000_000
X_eval = torch.zeros((n, 4))
# This should be O(n)
gp.posterior(X_eval)
```

Unless the above is run with ~12TiB of RAM, it will run out of memory as the covar matrix (size = 3 x 1000000^2) is evaluated, instead of only its diagonal.

## Expected Behavior

The above should run in O(n) and use <1GiB of RAM.

## System information

- `botorch==0.5.1`
- `gpytorch==1.5.1`
- `1.9.0+cu111`
- Win10 x64

## Additional context

The covar matrix is de-lazified in two places:
1. https://github.com/pytorch/botorch/blob/d0326cb2237cd2e769a059cc0f5006b5a9b3edda/botorch/models/gpytorch.py#L347 This can be changed to `lazy_covariance_matrix`, but since lazy tensors don't support `select`, line 352 needs to be changed as well to:
    ```
    covar_x[(slice(None),) * output_dim_idx + (t,)],
    ```
    _Note that this fix triggers [another bug](https://github.com/cornellius-gp/gpytorch/issues/1809) in GPyTorch so that needs to be fixed first._
2. https://github.com/cornellius-gp/gpytorch/blob/c96965c8aefac1874652fb44321541f5972eaa88/gpytorch/lazy/lazy_evaluated_kernel_tensor.py#L240 (called indirectly by `MultitaskMultivariateNormal.from_independent_mvns`).

**Comment by Balandat** [comment]
Thanks for flagging this; this is not great. Some context on this: gpytorch lazy tensor indexing is not particularly fast, so we had found in the past that often being non-lazy and indexing (rather than indexing the lazy directly) was actually faster up to medium size posteriors. Not sure if this is still the case or not. Either way, of course if you have millions of data points that is not true, so at the very least we should gate this behavior based on the matrix size here. 

I just put up https://github.com/cornellius-gp/gpytorch/pull/1811 to fix the "other bug".

Do you know whether this particular [`evaluate_kernel`](https://github.com/cornellius-gp/gpytorch/blob/c96965c8aefac1874652fb44321541f5972eaa88/gpytorch/lazy/lazy_evaluated_kernel_tensor.py#L240) call actually instantiates the matrix? Note that that function may also return a `LazyTensor` (see [here](https://github.com/cornellius-gp/gpytorch/blob/c96965c8aefac1874652fb44321541f5972eaa88/gpytorch/lazy/lazy_evaluated_kernel_tensor.py#L275-L276)).

**Comment by valtron** [comment]
Yeah, it instantiates the matrix [here](https://github.com/cornellius-gp/gpytorch/blob/c96965c8aefac1874652fb44321541f5972eaa88/gpytorch/kernels/kernel.py#L398) (due to https://github.com/cornellius-gp/gpytorch/blob/c96965c8aefac1874652fb44321541f5972eaa88/gpytorch/lazy/lazy_evaluated_kernel_tensor.py#L281).

> being non-lazy and indexing (rather than indexing the lazy directly) was actually faster up to medium size posteriors

That's true, I can see that being important. If posteriors become lazy by default, it should still be possible to use `settings.lazily_evaluate_kernels(False)` to get the old behaviour though.

**Comment by Balandat** [comment]
Cc @sdaulton

**Comment by sdaulton** [comment]
Another related inefficiency we aught to address is that if the outputs are independent, then we unnecessarily materialize a `(batch_shape x) nt x nt` root decomposition when posterior sampling with provided `base_samples`: https://github.com/cornellius-gp/gpytorch/blob/75638add97f13e7a29e3ed87f2924308c6950f14/gpytorch/distributions/multivariate_normal.py#L162. If we leverage the knowledge that the root decomposition is block diagonal, then we would only need to materialize a  `batch_shape x t x n x n` root decomposition

**Comment by valtron** [comment]
For [(2)](https://github.com/cornellius-gp/gpytorch/blob/c96965c8aefac1874652fb44321541f5972eaa88/gpytorch/lazy/lazy_evaluated_kernel_tensor.py#L240), changing
```diff
- return self.evaluate_kernel()._unsqueeze_batch(dim)
+ return self[(slice(None),) * dim + (None,)]
```
works.

**Comment by Balandat** [comment]
Nice! Are you willing to make a PR for gpytorch or should I just take care of that?

**Comment by valtron** [comment]
Sure, I can make the PR.

**Comment by sdaulton** [comment]
> > Another related inefficiency we aught to address is that if the outputs are independent, then we unnecessarily materialize a (batch_shape x) nt x nt root decomposition when posterior sampling with provided base_samples
> 
> Are we not doing this already?

I was mistaken. This is done correctly.

**Comment by valtron** [comment]
I created a PR for (2): https://github.com/cornellius-gp/gpytorch/pull/1813. I can also make a PR for (1), if that fix looks OK?

**Comment by Balandat** [comment]
I requested review from the experts for your gpytorch PR - messing with the indexing in the lazy tensors can get quite hairy so I'd like to have them take a look. Other than than I think your fix for (1) sounds great and we'd love a PR for it (if the gpytorch change goes in).

**Comment by valtron** [comment]
Closed by #976.
